### PR TITLE
Fix batched_nms for rotated box and add type hint for nms.py

### PR DESCRIPTION
**Pull request by liuyanyi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

When set `class_agnostic=False` in `batched_nms`, it will add an offset to det boxes, for horizional bbox in xyxy, the code works fine. But in rotated box (x_ctr, y_ctr, w, h, angle), only center point need offset, so I add a branch to solve that.

Add type hint for nms.py. #1948 

## Modification

- Update mmcv/ops/nms.py
- Update tests/test_ops/test_nms_rotated.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by zhouzaida** [comment]
please @zytx121 have a look.

**Comment by liuyanyi** [comment]
> Hi @liuyanyi , it would be better to add a unit test for this modification.
> 
> Unit tests can be added at
> 
> https://github.com/open-mmlab/mmcv/blob/70944a261d492c2b0b765ad009f78f3506ce0f16/tests/test_ops/test_nms_rotated.py#L73

Done

File: mmcv/ops/nms.py [hunk 1]
```diff
@@ -250,7 +264,12 @@ def soft_nms(boxes,
         return dets.to(device=boxes.device), inds.to(device=boxes.device)
 
 
-def batched_nms(boxes, scores, idxs, nms_cfg, class_agnostic=False):
+def batched_nms(
+        boxes: torch.Tensor,
+        scores: torch.Tensor,
+        idxs: torch.Tensor,
+        nms_cfg: Optional[Dict],
```
Comment (zytx121) [inline mmcv/ops/nms.py:271]: In docsting，`nms_cfg (dict | None): ` -> `nms_cfg (dict, optional): `

File: mmcv/ops/nms.py [hunk 2]
```diff
@@ -378,27 +408,32 @@ def nms_match(dets, iou_threshold):
...
-        dets (Tensor):  Rotated boxes in shape (N, 5). They are expected to
-            be in (x_ctr, y_ctr, width, height, angle_radian) format.
-        scores (Tensor): scores in shape (N, ).
+        dets (torch.Tensor):  Rotated boxes in shape (N, 5).
+            They are expected to be in
+            (x_ctr, y_ctr, width, height, angle_radian) format.
+        scores (torch.Tensor): scores in shape (N, ).
         iou_threshold (float): IoU thresh for NMS.
-        labels (Tensor): boxes' label in shape (N,).
+        labels (torch.Tensor): boxes' label in shape (N,).
```
Comment (zytx121) [inline mmcv/ops/nms.py:436]:
```suggestion
        labels (torch.Tensor, optional): boxes' label in shape (N,).
```

File: mmcv/ops/nms.py [hunk 3]
```diff
@@ -14,8 +15,9 @@
 class NMSop(torch.autograd.Function):
 
     @staticmethod
-    def forward(ctx, bboxes, scores, iou_threshold, offset, score_threshold,
-                max_num):
+    def forward(ctx: Any, bboxes: torch.Tensor, scores: torch.Tensor,
```
Comment (zytx121) [inline mmcv/ops/nms.py:18]: Maybe we can use ``from torch import Tensor``,  and use `Tensor` to replace `torch.Tensor` in Type hint.

File: mmcv/ops/nms.py (hunk 1 above)
Comment (liuyanyi) [reply mmcv/ops/nms.py]: fixed

**Review by zytx121 — APPROVED** [review]
LGMT

File: mmcv/ops/nms.py [hunk 4]
```diff
@@ -33,8 +35,8 @@ def forward(ctx, bboxes, scores, iou_threshold, offset, score_threshold,
         return inds
 
     @staticmethod
-    def symbolic(g, bboxes, scores, iou_threshold, offset, score_threshold,
-                 max_num):
+    def symbolic(g: Any, bboxes: Tensor, scores: Tensor, iou_threshold: float,
+                 offset: int, score_threshold: float, max_num: int) -> Tensor:
```
Comment (zhouzaida) [inline mmcv/ops/nms.py:39]:
```suggestion
    def symbolic(g, bboxes, scores, iou_threshold, offset, score_threshold,
                 max_num):
```

Comment (zhouzaida) [reply mmcv/ops/nms.py:39]: We do not need to add type hints for this method.

File: mmcv/ops/nms.py [hunk 5]
```diff
@@ -98,8 +101,9 @@ def forward(ctx, boxes, scores, iou_threshold, sigma, min_score, method,
         return dets, inds
 
     @staticmethod
-    def symbolic(g, boxes, scores, iou_threshold, sigma, min_score, method,
-                 offset):
+    def symbolic(g: Any, boxes: Tensor, scores: Tensor, iou_threshold: float,
+                 sigma: float, min_score: float, method: int,
+                 offset: int) -> Tuple[Tensor, Tensor]:
```
Comment (zhouzaida) [inline mmcv/ops/nms.py:106]: same as above

File: mmcv/ops/nms.py [hunk 6]
```diff
@@ -115,8 +118,16 @@ def symbolic(g, boxes, scores, iou_threshold, sigma, min_score, method,
         return nms_out
 
 
+nms_input_type = Union[Tensor, np.ndarray]
```
Comment (zhouzaida) [inline mmcv/ops/nms.py:121]: How about renaming `nms_input_type` to `array_like_type`?

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**